### ISSUE TITLE
Some ArrayTrie methods throw StackOverflowError when cointaining a very large entry

### ISSUE THREAD

**Issue by lorban** [eclipse/jetty.project#7517]
**Jetty version(s)**
9.4.x

**Description**
`keySet()` and `toString()` recurse too deep on large entries, so they're likely to throw `StackOverflowError`.

Consider the following:

```java
Trie<String> trie = new ArrayTrie<>(Character.MAX_VALUE - 1);

char[] c1 = new char[Character.MAX_VALUE - 1];
Arrays.fill(c1, 'a');
String huge = new String(c1);
assertTrue(trie.put(huge, "wow"));
assertThat(trie.get(huge), is("wow"));

assertThat(trie.keySet(), notNullValue());   // <--- throws StackOverflowError
assertThat(trie.toString(), notNullValue()); // <--- throws StackOverflowError
```

**Comment by joakime** [comment]
@lorban is this done/complete now?

**Comment by lorban** [comment]
@joakime yes, it's all done and merged across the necessary branches. Thanks for the heads up.

**Comment by gregw** [comment]
@lorban did these changes make it into the jetty-12 branch?   I think we need a process to track such issues?

**Comment by lorban** [comment]
@gregw no, they did not get merged to 12.0.x and you're right that we're going to need a process to remind us that.

I'm going to merge that right away.

**Comment by lorban** [comment]
Merged to 12.0.x.

**Comment by 4devwithgit** [comment]
Is this addressed in jetty 12.0.16? We are seeing the similar stack trace

**Comment by sbordet** [comment]
@4devwithgit please open a new issue with the exact stack trace and all the details.
